kostyor/upgrades/driver.py

```diff
@@ -0,0 +1,85 @@
+import abc
+import six
+
+
+@six.add_metaclass(abc.ABCMeta)
+class UpgradeDriver():
+
+    @abc.abstractproperty
+    def supports_upgrade_rollback(self):
+        """Property used to indicate if an implementation supports rolling back
+        an upgrade
+        """
+        pass
+
+    @abc.abstractmethod
+    def pre_host_upgrade_hook(self, host):
+        """Called by the decision engine before a host is upgraded,
+        allows an UpgradeDriver the opportunity to do any operations required
+        before the host is upgraded
+
+        Params
+
+        host: a kostyor.db.models.Host instance, representing the host that is
+        scheduled to be upgraded
+        """
+        pass
+
+    @abc.abstractmethod
+    def pre_service_upgrade_hook(self, service):
+        """Called by the decision engine before a service is upgraded,
+        allows an UpgradeDriver the opportunity to do any operations required
+        before a service is upgraded
+
+        Params
+
+        service: a kostyor.db.models.Service instance, representing the service
+        that is scheduled to be upgraded
+        """
+        pass
+
+    @abc.abstractmethod
+    def post_host_upgrade_hook(self, host):
+        """Called by the decision engine after a host is upgraded,
+        allows an UpgradeDriver the opportunity to do any operations required
+        after the host is upgraded
+
+        Params
+
+        host: a kostyor.db.models.Host instance, representing the host that is
+        scheduled to be upgraded
+        """
+        pass
+
+    @abc.abstractmethod
+    def post_service_upgrade_hook(self, service):
+        """Called by the decision engine after a service is upgraded,
+        allows an UpgradeDriver the opportunity to do any operations required
+        after a service is upgraded
+
+        Params
+
+        service: a kostyor.db.models.Service instance, representing the service
+        that is scheduled to be upgraded
+        """
+        pass
+
+    @abc.abstractmethod
+    def cancel_upgrade(self):
```

Comment:
> My concern about defining an `__init__` method is that once we set that in the signature, we're pretty much stuck with that forever, and it's a very important method.

Right. That's why I think `__init__` **should not** be a part of interface (abstractmethod). It should be provided through base class. The idea is, we control how to create driver instances and which arguments to pass. Driver implementations, generally, should not override `__init__` and should not care about its construction at all. They just need to provide implementations for abstract methods.

> One thing to think on: At least with defining all the upgrade methods, we can define methods with different sets of args for backwards compatibility.

I'm sorry, I didn't get it. :( Could you please explain? Python doesn't have method overloading by args.